File: Data Structures & Algorithms/palindromic-substrings/submission-6.cpp
class Solution {
public:
    int countSubstrings(string s) {
        std::string t{"#"};
        for (char c : s) {
            t += c;
            t += "#";
        }
        int l{}, r{};
        std::vector<int> p(t.size(), 0);
        for (int i{}; i < t.size(); ++i) {
            if (i <  r) {
                p[i] = std::min(r - i, p[l + r - i]);
            }
            else p[i] = 0;
            while (i + p[i] + 1 < t.size() && i - p[i] - 1 >=0 && t[i + p[i] + 1] == t[i - p[i] - 1]) {
                p[i] ++;
            } 
            if (i + p[i] > r) {
                l = i - p[i];
                r = i + p[i];
            }
        }
        int res{};
        for (int i{}; i < t.size(); ++i) {
            res += (p[i] + 1) / 2;
        }
        return res;
    }
};

File: Data Structures & Algorithms/partition-labels/submission-6.java
class Solution {
    public List<Integer> partitionLabels(String s) {
        HashMap<Character, Integer> last = new HashMap<>();
        int i =0;
        for (char c : s.toCharArray()) {
            last.put(c, i);
            i ++;
        }
        int l = 0;
        int r = 0;
        ArrayList<Integer> res = new ArrayList<>();
        while (l < s.length()) {
            for (i= l; i<= r; ++i) {
                r = Math.max(last.get(s.charAt(i)), r);

            }
            res.add(r - l + 1);
            l = r + 1;
            r = l;
        }   
        return res;
    }
}
